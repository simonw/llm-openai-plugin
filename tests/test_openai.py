import json
import llm
import os
from pydantic import BaseModel
import pytest

API_KEY = os.environ.get("PYTEST_OPENAI_API_KEY", None) or "badkey"


def test_plugin_is_installed():
    model_ids = [model.model_id for model in llm.get_models()]
    assert "openai/gpt-4o-mini" in model_ids


@pytest.mark.parametrize(
    "options",
    (
        {"max_output_tokens": 24},
        {"temperature": 0.5},
        {"top_p": 0.5},
        {"store": True},
        {"truncation": "auto"},
    ),
)
@pytest.mark.vcr
def test_options(options, snapshot, vcr):
    model = llm.get_model("openai/gpt-4o-mini")
    response = model.prompt("say hi", key=API_KEY, stream=False, **options)
    assert response.text() == snapshot
    # Was the option sent to the API?
    api_input = json.loads(vcr.requests[0].body)
    assert all(item in api_input.items() for item in options.items())
    usage = response.usage()
    assert usage.input == 27


@pytest.mark.asyncio
@pytest.mark.vcr
async def test_async_model(snapshot):
    model = llm.get_async_model("openai/gpt-4o-mini")
    response = await model.prompt("say hi", key=API_KEY)
    output = await response.text()
    assert output == snapshot
    usage = await response.usage()
    assert usage.input == 27
    assert usage.output == 11


class Dog(BaseModel):
    name: str
    age: int
    bio: str


@pytest.mark.asyncio
@pytest.mark.vcr
async def test_async_model_schema(snapshot):
    model = llm.get_async_model("openai/gpt-4o-mini")
    response = await model.prompt("invent a dog", key=API_KEY, schema=Dog)
    output = await response.text()
    assert json.loads(output) == snapshot


@pytest.mark.vcr
def test_tools(snapshot):
    model = llm.get_model("openai/gpt-5-mini")

    def simple_tool(number):
        "A simple tool"
        return "This is a simple tool, {}".format(number)

    chain_response = model.chain(
        "Call simple_tool passing 5",
        tools=[simple_tool],
        key=API_KEY
    )
    output = chain_response.text()
    assert output == snapshot
