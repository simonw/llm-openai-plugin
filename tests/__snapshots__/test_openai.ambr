# serializer version: 1
# name: test_async_model
  'Hi there! How can I assist you today?'
# ---
# name: test_async_model_schema
  dict({
    'age': 5,
    'bio': 'Barkley is a playful and friendly Golden Retriever mix with a love for adventure. He enjoys playing fetch at the park, splashing in lakes, and cuddling on the couch after a long day of exploring. With a fluffy coat and a wagging tail, Barkley brings joy to everyone he meets.',
    'name': 'Barkley',
  })
# ---
# name: test_options[options0]
  'Hi there! How can I assist you today?'
# ---
# name: test_options[options1]
  'Hi there! How can I assist you today?'
# ---
# name: test_options[options2]
  'Hi there! How can I assist you today?'
# ---
# name: test_options[options3]
  'Hi there! How can I assist you today?'
# ---
# name: test_options[options4]
  'Hi there! How can I assist you today?'
# ---
# name: test_tools
  '''
  I called simple_tool with 5; it returned:
  "This is a simple tool, 5"
  
  Anything else youâ€™d like me to run or change?
  '''
# ---
